invoice/invoice.go

```diff
@@ -0,0 +1,938 @@
+	// Optional.
+	RoutingInfo []ExtraRoutingInfo
+}
+
+// ExtraRoutingInfo holds the information needed to route a payment along one
+// private channel.
+type ExtraRoutingInfo struct {
+	// PubKey is the public key of the node at the start of this channel.
+	PubKey *btcec.PublicKey
+
+	// ShortChanID is the channel ID of the channel.
+	ShortChanID uint64
+
+	// Fee is the fee required for routing along this channel.
+	Fee uint64
+
+	// CltvExpDelta is this channel's cltv expiry delta.
+	CltvExpDelta uint16
+}
+
+// Amount is a functional option that allows callers of NewInvoice to set the
+// amount in millisatoshis that the Invoice should encode.
+func Amount(milliSat lnwire.MilliSatoshi) func(*Invoice) {
+	return func(i *Invoice) {
+		i.MilliSat = &milliSat
+	}
+}
+
+// Destination is a functional option that allows callers of NewInvoice to
+// explicitly set the pubkey of the Invoice's destination node.
+func Destination(destination *btcec.PublicKey) func(*Invoice) {
+	return func(i *Invoice) {
+		i.Destination = destination
+	}
+}
+
+// Description is a functional option that allows callers of NewInvoice to set
+// the payment description of the created Invoice.
+// Note: Must be used if and only if DescriptionHash is not used.
+func Description(description string) func(*Invoice) {
+	return func(i *Invoice) {
+		i.Description = &description
+	}
+}
+
+// DescriptionHash is a functional option that allows callers of NewInvoice to
+// set the payment description hash of the created Invoice.
+// Note: Must be used if and only if Description is not used.
+func DescriptionHash(descriptionHash [32]byte) func(*Invoice) {
+	return func(i *Invoice) {
+		i.DescriptionHash = &descriptionHash
+	}
+}
+
+// Expiry is a functional option that allows callers of NewInvoice to set the
+// expiry of the created Invoice. If not set, a default expiry of 60 min will
+// be implied.
+func Expiry(expiry time.Time) func(*Invoice) {
+	return func(i *Invoice) {
+		i.Expiry = &expiry
+	}
+}
+
+// FallbackAddr is a functional option that allows callers of NewInvoice to set
+// the Invoice's fallback on-chain address that can be used for payment in case
+// the Lightning payment fails
+func FallbackAddr(fallbackAddr btcutil.Address) func(*Invoice) {
+	return func(i *Invoice) {
+		i.FallbackAddr = fallbackAddr
+	}
+}
+
+// RoutingInfo is a functional option that allows callers of NewInvoice to set
+// one or more entries containing extra routing information for a private route
+// to the target node.
+func RoutingInfo(routingInfo []ExtraRoutingInfo) func(*Invoice) {
+	return func(i *Invoice) {
+		i.RoutingInfo = routingInfo
+	}
+}
+
+// NewInvoice creates a new Invoice object. The last parameter is a set of
+// variadic argumements for setting optional fields of the invoice.
+// Note: Either Description  or DescriptionHash must be provided for the Invoice
+// to be considered valid.
+func NewInvoice(net *chaincfg.Params, paymentHash [32]byte,
```

Comment:
Very good point! But will be changing the package name to `zpay32` in a follow-up PR. Then it might make sense to keep the `zpay32.NewInvoice` name..?